File: frontend/src/types/item.ts
/**
 * 아이템(업무) 타입 정의
 */

import type { PageResponse } from './api'

export type ItemStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'PENDING' | 'COMPLETED' | 'DELETED'
export type Priority = 'URGENT' | 'HIGH' | 'NORMAL' | 'LOW'

export interface Item {
  itemId: number
  boardId: number
  boardName?: string
  groupId?: number
  groupName?: string
  groupColor?: string
  categoryId?: number
  categoryName?: string
  categoryColor?: string
  title: string
  content?: string
  description?: string  // 상세 내용 (마크다운)
  status: ItemStatus
  priority: Priority
  assigneeUsername?: string   // 담당자 USERNAME (백엔드 FK 참조용)
  assigneeName?: string       // 담당자 표시용 이름
  requestDate?: string        // 요청일
  dueDate?: string            // 마감일 (지연 판정에 사용)
  sortOrder?: number
  completedAt?: string    // 완료일시
  deletedAt?: string      // 삭제일시
  commentCount?: number
  createdAt: string
  createdBy?: string          // 생성자 USERNAME
  createdByName?: string
  updatedAt?: string
  updatedByName?: string
  properties?: ItemProperty[]
  propertyValues?: Record<number, unknown>
  // 공유/이관 정보
  isSharedToMe?: boolean        // 공유받은 업무 여부
  sharedByUsername?: string     // 공유해준 사용자 USERNAME
  sharedByUserName?: string     // 공유해준 사용자 이름
  transferredFrom?: number      // 이관 원본 보드 ID
  transferredByUsername?: string // 이관해준 사용자 USERNAME
  transferredByUserName?: string // 이관해준 사용자 이름
  // 배당 정보 (v2.1)
  isAssignedToMe?: boolean       // 배당받은 업무 여부
  assignedByUsername?: string    // 배당해준 사용자 USERNAME
  assignedByUserName?: string    // 배당해준 사용자 이름
  assignedAt?: string            // 배당일시
  // 배당 대상 정보 (소유자 화면용 v2.1)
  assignedToUsername?: string    // 배당 대상 사용자 USERNAME
  assignedToUserName?: string    // 배당 대상 사용자 이름
}

export interface ItemProperty {
  itemPropertyId: number
  itemId: number
  propertyId: number
  propertyName?: string
  propertyCode?: string
  propertyType?: string
  ownerType?: 'GLOBAL' | 'MANAGER' | 'USER'  // 속성 소유자 타입
  dataSourceType?: 'INTERNAL' | 'EXTERNAL'   // 데이터 소스 타입
  externalQueryId?: number                    // 외부 쿼리 ID
  value?: unknown           // 실제 값
  displayValue?: string     // 표시용 값 (옵션명, 사용자명 등)
  color?: string            // 옵션 색상
  sortOrder?: number        // 정렬 순서 (TB_ITEM_PROPERTY)
}

export interface ItemCreateRequest {
  title: string
  content?: string
  description?: string  // 상세 내용 (마크다운)
  status?: ItemStatus
  priority?: Priority
  groupId?: number
  categoryId?: number
  assigneeUsername?: string   // 담당자 USERNAME
  requestDate?: string        // 요청일
  dueDate?: string            // 마감일
  sortOrder?: number
  propertyIds?: number[]      // v2.0: 선택된 속성 ID 목록
  properties?: Record<number, unknown>
  propertySortOrders?: Record<number, number>  // 속성 정렬 순서 (propertyId -> sortOrder)
}

export interface ItemUpdateRequest {
  title?: string
  content?: string
  description?: string  // 상세 내용 (마크다운)
  status?: ItemStatus
  priority?: Priority
  groupId?: number
  categoryId?: number
  assigneeUsername?: string   // 담당자 USERNAME
  requestDate?: string        // 요청일
  dueDate?: string            // 마감일
  sortOrder?: number
  propertyIds?: number[]      // v2.0: 선택된 속성 ID 목록
  properties?: Record<number, unknown>
  propertySortOrders?: Record<number, number>  // 속성 정렬 순서 (propertyId -> sortOrder)
  /** null로 설정할 필드 목록 (예: ['requestDate', 'dueDate']) - snake_case로 전송됨 */
  clear_fields?: string[]
}

export interface ItemSearchRequest {
  keyword?: string
  status?: ItemStatus
  priority?: Priority
  assigneeUsername?: string   // 담당자 USERNAME
  groupId?: number
  departmentCode?: string     // 부서 CODE (외부 연동 키 정책)
  startDate?: string
  endDate?: string
  includeCompleted?: boolean
  includeDeleted?: boolean
  page?: number
  size?: number
  sortField?: string          // 정렬 필드 (sortOrder, createdAt, dueDate 등)
  sortDirection?: 'asc' | 'desc'  // 정렬 방향
}

/**
 * Cross-board 검색 요청 DTO
 */
export interface CrossBoardSearchRequest {
  keyword?: string
  status?: ItemStatus
  priority?: Priority
  assigneeUsername?: string   // 담당자 USERNAME
  groupId?: number
  departmentCode?: string     // 부서 CODE (외부 연동 키 정책)
  boardId?: number
  overdueOnly?: boolean
  startDate?: string
  endDate?: string
  page?: number
  size?: number
  sort?: string
}

/**
 * 페이징 응답 DTO (PageResponse<Item>의 타입 별칭)
 */
export type ItemPageResponse = PageResponse<Item>

/**
 * Cross-board 통계 DTO
 */
export interface CrossBoardStats {
  overdueCount: number
  pendingCount: number
  activeCount: number
  urgentOverdueCount: number
  highOverdueCount: number
}

/**
 * 업무 이관 요청 DTO
 */
export interface ItemTransferRequest {
  targetBoardId?: number
  targetUsername?: string   // 이관 대상 사용자 USERNAME
  reason?: string
}

/**
 * 업무 공유 정보 DTO
 */
export interface ItemShare {
  itemShareId: number
  itemId: number
  username: string          // 공유 사용자 USERNAME
  userName?: string         // 표시용 이름
  departmentName?: string
  permission: 'VIEW' | 'EDIT' | 'FULL'
  canView: boolean
  canEdit: boolean
  canDelete: boolean
  createdAt?: string
  updatedAt?: string
}

/**
 * 아이템 속성 순서 변경 요청 DTO
 */
export interface ItemPropertySortRequest {
  orders: PropertySortOrder[]
}

/**
 * 속성 순서 정보
 */
export interface PropertySortOrder {
  itemPropertyId: number
  sortOrder: number
}

/**
 * 아이템 순서 변경 요청 DTO
 */
export interface ItemReorderRequest {
  itemId: number
  newOrder: number
}

File: backend/src/main/java/com/taskflow/domain/Item.java
package com.taskflow.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 업무 아이템 엔티티
 *
 * 테이블: TB_ITEM
 *
 * USERNAME 기반 FK 참조 시스템:
 * - ASSIGNEE_USERNAME: 담당자 USERNAME 참조
 * - COMPLETED_BY, DELETED_BY, CREATED_BY, UPDATED_BY: USERNAME 참조
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Item {

    /**
     * 아이템 ID (PK)
     */
    private Long itemId;

    /**
     * 보드 ID (FK)
     */
    private Long boardId;

    /**
     * 그룹 ID (FK) - TB_GROUP 연동
     */
    private Long groupId;

    /**
     * 카테고리 ID (FK) - TB_PROPERTY_OPTION 연동
     */
    private Long categoryId;

    /**
     * 아이템 제목 (DB: CONTENT)
     */
    private String title;

    /**
     * 제목 원본 (DB: CONTENT) - title과 동일, 호환성용
     */
    private String content;

    /**
     * 상세 내용 - 마크다운 (DB: DESCRIPTION)
     */
    private String description;

    /**
     * 상태 (NOT_STARTED, IN_PROGRESS, PENDING, COMPLETED, DELETED)
     */
    private String status;

    /**
     * 우선순위 (URGENT, HIGH, NORMAL, LOW)
     */
    private String priority;

    /**
     * 담당자 USERNAME (FK → TB_USER.USERNAME)
     */
    private String assigneeUsername;

    /**
     * 요청일 (DB: REQUEST_DATE)
     */
    private LocalDate requestDate;

    /**
     * 마감일 (DB: DUE_DATE)
     */
    private LocalDate dueDate;

    /**
     * 이전 상태 (삭제/복원 전 상태)
     */
    private String previousStatus;

    /**
     * 완료 처리자 USERNAME
     */
    private String completedBy;

    /**
     * 삭제일시
     */
    private LocalDateTime deletedAt;

    /**
     * 삭제자 USERNAME
     */
    private String deletedBy;

    /**
     * 이관 원본 보드 ID
     */
    private Long transferredFrom;

    /**
     * 이관 일시
     */
    private LocalDateTime transferredAt;

    /**
     * 정렬 순서
     */
    private Integer sortOrder;

    /**
     * 생성일시
     */
    private LocalDateTime createdAt;

    /**
     * 생성자 USERNAME
     */
    private String createdBy;

    /**
     * 수정일시
     */
    private LocalDateTime updatedAt;

    /**
     * 수정자 USERNAME
     */
    private String updatedBy;

    // =============================================
    // 추가 필드 (Mapper에서 JOIN으로 설정)
    // =============================================

    /**
     * 보드명
     */
    private String boardName;

    /**
     * 그룹명
     */
    private String groupName;

    /**
     * 그룹 색상
     */
    private String groupColor;

    /**
     * 카테고리명 (조인)
     */
    private String categoryName;

    /**
     * 카테고리 색상 (조인)
     */
    private String categoryColor;

    /**
     * 담당자명 (조인)
     */
    private String assigneeName;

    /**
     * 생성자명 (조인)
     */
    private String createdByName;

    /**
     * 수정자명 (조인)
     */
    private String updatedByName;

    /**
     * 댓글 수
     */
    private Integer commentCount;

    // =============================================
    // 공유/이관 정보 (Mapper에서 JOIN으로 설정)
    // =============================================

    /**
     * 공유받은 업무 여부 (현재 사용자 기준)
     */
    private Boolean isSharedToMe;

    /**
     * 공유해준 사용자 USERNAME
     */
    private String sharedByUsername;

    /**
     * 공유해준 사용자 이름
     */
    private String sharedByUserName;

    /**
     * 이관해준 사용자 USERNAME (조인)
     */
    private String transferredByUsername;

    /**
     * 이관해준 사용자 이름 (조인)
     */
    private String transferredByUserName;

    /**
     * 동적 속성값 목록
     */
    @Builder.Default
    private List<ItemProperty> properties = new ArrayList<>();

    /**
     * 동적 속성값 맵 (propertyId -> value)
     */
    @Builder.Default
    private Map<Long, Object> propertyValues = new HashMap<>();

    // =============================================
    // 상수: 상태
    // =============================================

    public static final String STATUS_NOT_STARTED = "NOT_STARTED";
    public static final String STATUS_IN_PROGRESS = "IN_PROGRESS";
    public static final String STATUS_COMPLETED = "COMPLETED";
    public static final String STATUS_DELETED = "DELETED";
    public static final String STATUS_PENDING = "PENDING";

    // =============================================
    // 상수: 우선순위
    // =============================================

    public static final String PRIORITY_URGENT = "URGENT";
    public static final String PRIORITY_HIGH = "HIGH";
    public static final String PRIORITY_NORMAL = "NORMAL";
    public static final String PRIORITY_LOW = "LOW";

    // =============================================
    // 편의 메서드
    // =============================================

    /**
     * 완료 상태 여부
     */
    public boolean isCompleted() {
        return STATUS_COMPLETED.equals(status);
    }

    /**
     * 삭제 상태 여부
     */
    public boolean isDeleted() {
        return STATUS_DELETED.equals(status);
    }

    /**
     * 진행 중 여부
     */
    public boolean isInProgress() {
        return STATUS_IN_PROGRESS.equals(status);
    }

    /**
     * 시작 전 여부
     */
    public boolean isNotStarted() {
        return STATUS_NOT_STARTED.equals(status);
    }

    /**
     * 활성 아이템 여부 (완료/삭제 아님)
     */
    public boolean isActive() {
        return !isCompleted() && !isDeleted();
    }

    /**
     * 속성값 추가
     */
    public void addProperty(ItemProperty property) {
        if (properties == null) {
            properties = new ArrayList<>();
        }
        properties.add(property);
    }

    /**
     * 속성값 맵에 추가
     */
    public void putPropertyValue(Long propertyId, Object value) {
        if (propertyValues == null) {
            propertyValues = new HashMap<>();
        }
        propertyValues.put(propertyId, value);
    }

    // =============================================
    // 명시적 Getter (Lombok 빌드 호환성)
    // =============================================

    /**
     * 공유해준 사용자 USERNAME getter
     */
    public String getSharedByUsername() {
        return sharedByUsername;
    }

    /**
     * 공유해준 사용자 이름 getter
     */
    public String getSharedByUserName() {
        return sharedByUserName;
    }

    /**
     * 이관해준 사용자 USERNAME getter
     */
    public String getTransferredByUsername() {
        return transferredByUsername;
    }

    /**
     * 이관해준 사용자 이름 getter
     */
    public String getTransferredByUserName() {
        return transferredByUserName;
    }

    /**
     * 정렬 순서 getter
     */
    public Integer getSortOrder() {
        return sortOrder;
    }

    /**
     * 정렬 순서 setter
     */
    public void setSortOrder(Integer sortOrder) {
        this.sortOrder = sortOrder;
    }
}
